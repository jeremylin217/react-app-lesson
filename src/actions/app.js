import { createAction } from 'redux-actions'
import { APP_CLICK } from '@/constants/actionTypes'

export const click = createAction(APP_CLICK)
