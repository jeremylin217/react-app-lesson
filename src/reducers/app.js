import { handleActions } from 'redux-actions'

const initState = {
  count: 0,
}

export default handleActions(
  {
    APP_CLICK(state, action) {
      return {
        ...state,
        count: state.count + 1,
      }
    },
  },
  initState
)
