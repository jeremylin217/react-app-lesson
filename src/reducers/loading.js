export const createLoadingSelector = actions => state =>
  actions.some(action => state.loading[action])

const loadingReducer = (state = {}, action) => {
  const { type } = action
  const matches = /(.*)_(REQUEST|RECEIVE|FAILURE)/.exec(type)

  // not a *_REQUEST / *_SUCCESS / *_FAILURE actions, so we ignore them
  if (!matches) return state

  const [, requestName, requestState] = matches
  return {
    ...state,
    // Store whether a request is happening at the moment or not
    // e.g. will be true when receiving GET_TODOS_REQUEST
    //      and false when receiving GET_TODOS_RECEIVE / GET_TODOS_FAILURE
    [requestName]: requestState === 'REQUEST',
  }
}

export default loadingReducer
