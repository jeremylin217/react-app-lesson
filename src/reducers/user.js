import { handleActions } from 'redux-actions'

const initState = {
  id: null,
  username: '',
  name: 'Jeremy',
  email: '',
}

export default handleActions(
  {
    GET_USER_REQUEST(state, action) {
      return initState
    },
    GET_USER_RECEIVE(state, action) {
      return action.payload.user
    },
  },
  initState
)
