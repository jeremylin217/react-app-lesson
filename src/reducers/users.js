import { handleActions } from 'redux-actions'

const initState = []

export default handleActions(
  {
    GET_USERS_REQUEST(state, action) {
      return initState
    },
    GET_USERS_RECEIVE(state, action) {
      return action.payload.users
    },
  },
  initState
)
