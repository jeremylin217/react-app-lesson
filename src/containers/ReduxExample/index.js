import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'
import { bindActionCreators } from 'redux'
import Counter from '@/components/Counter'
import * as appActions from '@/actions/app'

const mapStateToProps = state => ({
  count: state.app.count,
})

const mapDispatchToProps = dispatch => ({
  appActions: bindActionCreators(appActions, dispatch),
  // oneAction: (...args) => dispatch(oneAction(...args))
})

class ReduxExample extends Component {
  state = {
    color: '#000000',
  }

  colorChangeHandler = e => {
    this.setState({
      color:
        '#' + (0x1000000 + Math.random() * 0xffffff).toString(16).substr(1, 6),
    })
  }

  constructor(props) {
    super(props)
    console.log('constructor')
  }

  // shouldComponentUpdate(nextProps, nextState) {
  //   return false
  // }

  componentWillMount() {
    console.log('componentWillMount')
  }

  componentDidMount() {
    console.log('componentDidMount')
  }

  componentDidUpdate() {
    console.log('componentDidUpdate')
  }

  componentWillUnmount() {
    console.log('componentWillUnmount')
  }

  render() {
    const { color } = this.state
    const { count, appActions } = this.props
    return (
      <div className="Page">
        <div className="container text-center">
          <h1>Redux Example</h1>
          <hr />
          <Counter count={count} textColor={color} />
          <button className="btn btn-primary" onClick={appActions.click}>
            Click Me
          </button>
          &nbsp;
          <button
            className="btn btn-secondary"
            onClick={this.colorChangeHandler}
          >
            Change Color
          </button>
          <hr />
          <Link to="/">Back to Home</Link>
        </div>
      </div>
    )
  }
}

ReduxExample.propTypes = {
  count: PropTypes.number.isRequired,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(ReduxExample)
