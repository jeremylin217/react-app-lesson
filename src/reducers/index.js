import { combineReducers } from 'redux'
import { connectRouter } from 'connected-react-router'

import app from './app'
import user from './user'

export default history =>
  combineReducers({
    router: connectRouter(history),
    app,
    user,
  })
