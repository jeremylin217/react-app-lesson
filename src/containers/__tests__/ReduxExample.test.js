import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import configureStore from 'redux-mock-store'
import ReduxExample, {
  ReduxExample as ReduxExampleComponent,
} from '@/containers/ReduxExample'
import { MemoryRouter } from 'react-router'
import { initState as app } from '@/reducers/app'
import { Provider } from 'react-redux'

const middlewares = []
const mockStore = configureStore(middlewares)

describe('<ReduxExample />', () => {
  let wrapper
  let onClick

  beforeEach(() => {
    onClick = jest.fn()
    wrapper = mount(
      <MemoryRouter>
        <ReduxExampleComponent count={0} appActions={{ click: onClick }} />
      </MemoryRouter>
    )
  })

  it('should render', () => {
    expect(toJson(wrapper)).toMatchSnapshot()
  })

  it('should render "Click Me" button', () => {
    const clickMeButton = wrapper.find('button#clickMeButton').first()
    expect(clickMeButton).toBeDefined()
  })

  it('"Click Me" button click calls onClick', () => {
    const clickMeButton = wrapper.find('button#clickMeButton').first()
    // clickMeButton.simulate('click')
    clickMeButton.props().onClick()
    expect(onClick).toBeCalled()
  })
})

describe('<ReduxExample /> with redux store', () => {
  let wrapper

  beforeEach(() => {
    const initialState = { app }
    const store = mockStore(initialState)
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter>
          <ReduxExample />
        </MemoryRouter>
      </Provider>
    )
  })
})
