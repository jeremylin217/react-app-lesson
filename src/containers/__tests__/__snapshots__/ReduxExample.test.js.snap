// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReduxExample /> should render 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "test",
        "pathname": "/example/redux",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "test",
            "pathname": "/example/redux",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "test",
          "pathname": "/example/redux",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <ReduxExample
      appActions={
        Object {
          "click": [MockFunction],
        }
      }
      count={0}
    >
      <div
        className="Page"
      >
        <div
          className="container text-center"
        >
          <h1>
            Redux Example
          </h1>
          <hr />
          <Counter
            count={0}
            textColor="#000000"
          >
            <div>
              <div>
                Clicks
              </div>
              <styled.div
                count={0}
                textColor="#000000"
              >
                <StyledComponent
                  count={0}
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-bdVaJa",
                        "isStatic": false,
                        "lastClassName": "dZcQLs",
                        "rules": Array [
                          "
  margin-bottom: 1rem;
  font-weight: bold;
  font-size: ",
                          [Function],
                          "rem;
  color: ",
                          [Function],
                          ";
",
                        ],
                      },
                      "displayName": "styled.div",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-bdVaJa",
                      "target": "div",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  textColor="#000000"
                >
                  <div
                    className="sc-bdVaJa dZcQLs"
                  >
                    0
                  </div>
                </StyledComponent>
              </styled.div>
            </div>
          </Counter>
          <button
            className="btn btn-primary"
            id="clickMeButton"
            onClick={[MockFunction]}
          >
            Click Me
          </button>
           
          <button
            className="btn btn-secondary"
            onClick={[Function]}
          >
            Change Color
          </button>
          <hr />
          <Link
            to="/"
          >
            <a
              href="/"
              onClick={[Function]}
            >
              Back to Home
            </a>
          </Link>
        </div>
      </div>
    </ReduxExample>
  </Router>
</MemoryRouter>
`;

exports[`<ReduxExample /> with redux store should render 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter
    initialEntries={
      Array [
        Object {
          "key": "test",
          "pathname": "/example/redux",
        },
      ]
    }
  >
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "test",
              "pathname": "/example/redux",
              "search": "",
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "test",
            "pathname": "/example/redux",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Connect(ReduxExample)>
        <ReduxExample
          appActions={
            Object {
              "click": [Function],
            }
          }
          count={0}
        >
          <div
            className="Page"
          >
            <div
              className="container text-center"
            >
              <h1>
                Redux Example
              </h1>
              <hr />
              <Counter
                count={0}
                textColor="#000000"
              >
                <div>
                  <div>
                    Clicks
                  </div>
                  <styled.div
                    count={0}
                    textColor="#000000"
                  >
                    <StyledComponent
                      count={0}
                      forwardedComponent={
                        Object {
                          "$$typeof": Symbol(react.forward_ref),
                          "attrs": Array [],
                          "componentStyle": ComponentStyle {
                            "componentId": "sc-bdVaJa",
                            "isStatic": false,
                            "lastClassName": "dZcQLs",
                            "rules": Array [
                              "
  margin-bottom: 1rem;
  font-weight: bold;
  font-size: ",
                              [Function],
                              "rem;
  color: ",
                              [Function],
                              ";
",
                            ],
                          },
                          "displayName": "styled.div",
                          "foldedComponentIds": Array [],
                          "render": [Function],
                          "styledComponentId": "sc-bdVaJa",
                          "target": "div",
                          "toString": [Function],
                          "warnTooManyClasses": [Function],
                          "withComponent": [Function],
                        }
                      }
                      forwardedRef={null}
                      textColor="#000000"
                    >
                      <div
                        className="sc-bdVaJa dZcQLs"
                      >
                        0
                      </div>
                    </StyledComponent>
                  </styled.div>
                </div>
              </Counter>
              <button
                className="btn btn-primary"
                id="clickMeButton"
                onClick={[Function]}
              >
                Click Me
              </button>
               
              <button
                className="btn btn-secondary"
                onClick={[Function]}
              >
                Change Color
              </button>
              <hr />
              <Link
                to="/"
              >
                <a
                  href="/"
                  onClick={[Function]}
                >
                  Back to Home
                </a>
              </Link>
            </div>
          </div>
        </ReduxExample>
      </Connect(ReduxExample)>
    </Router>
  </MemoryRouter>
</Provider>
`;
